src/home/homework.c

```diff
@@ -1811,16 +1834,27 @@ static int home_lock(UserRecord *h) {
         if (r != USER_TEST_MOUNTED)
                 return log_error_errno(SYNTHETIC_ERRNO(ENOEXEC), "Home directory of %s is not mounted, can't lock.", h->user_name);
 
+        r = user_session_freezer(h->uid, true, &freezer);
+        if (r < 0)
+                log_warning_errno(r, "Failed to freeze user session, ignoring: %m");
+        else if (r == 0)
+                log_info("User session freeze disabled, skipping.");
+        else
+                log_info("Froze user session.");
+
         r = home_lock_luks(h, &setup);
         if (r < 0)
                 return r;
 
+        unit_freezer_cancel(&freezer); /* Don't thaw the user session. */
```

Comment:
And by having more specific functions you can avoid the confusing "cancel" logic here, and just use _done instead of _thaw_done for cleanup.

But maybe it's better to store the UnitFreezer object in memory rather than recreating them when unlocking...

Reply:
Can't do. I recreate them in a completely separate process

Reply:
(I could have sworn I replied to this last night, but I might not have pressed the submit button :shrug:)

> But maybe it's better to store the UnitFreezer object in memory rather than recreating them when unlocking

We cannot do this because the object is recreated in a completely separate process